Multiple networks

jackhodgkiss opened the pull request:
Add support for multiple networks.

jackhodgkiss committed:
feat: add internal and storage network

jackhodgkiss committed:
feat: vxlan vars in separate file and used by network-interfaces

jackhodgkiss committed:
feat: adapt vxlan vars to use `eno` or `eth0`

Also remove `configure-vxlan` as a post hook.

jackhodgkiss committed:
feat: define the various networks

This commit includes changes to how the network is split up into various
components.

jackhodgkiss committed:
feat: configure neutron for external network

The external network has been removed from kayobe and is instead handled
by neutron. Also the `out-of-band` network has been removed. Finally the
nova compute virt type has been changed back to the default.

jackhodgkiss committed:
Merge branch 'multinode-env' into multiple-networks

markgoddard commented on etc/kayobe/cephadm.yml:
```diff
@@ -23,19 +23,12 @@ cephadm_registry_username: "{{ stackhpc_docker_registry_username }}"
 cephadm_registry_password: "{{ stackhpc_docker_registry_password }}"
 
-# Cephadm admin network interface.
-cephadm_admin_interface: "{{ admin_oc_net_name | net_interface }}"
```
Why remove this?

jackhodgkiss replied:
Added back. Must have been a mistake when configuring the file.

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/all/vxlan.yml:
```diff
@@ -0,0 +1,6 @@
+---
```
This could be more helpful to users.

* Provide a comment about what to change and why
* Add the `vxlan_vni` variable that generally needs to be changed
* Use `vxlan_vni` in the interface name and group IP last octet to allow only changing one thing

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/all/vxlan.yml:
```diff
@@ -0,0 +1,6 @@
+---
+vxlan_phys_dev: "eno1"
```
It's a shame this doesn't work the normal way where the admin interface goes in group_vars/<group> and is referenced here as `admin_oc_interface`.

markgoddard commented on etc/kayobe/hooks/overcloud-host-configure/post.d/50-configure-vxlan.yml:
```diff
@@ -1 +0,0 @@
-../../../ansible/configure-vxlan.yml
```
Docs need to be updated to add it as a pre hook.

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/storage/vxlan.yml:
```diff
@@ -0,0 +1,6 @@
+---
+vxlan_phys_dev: "eth0"
```
Only this variable needs to be overridden. Ideally though we would just reference `admin_oc_interface` in group_vars/all/vxlan

markgoddard commented on etc/kayobe/environments/ci-multinode/networks.yml:
```diff
@@ -8,87 +8,124 @@
 # Name of the network used for admin access to the overcloud
 #admin_oc_net_name:
-admin_oc_net_name: admin
+admin_oc_net_name: admin_oc
```
Changing the name of this network means that the terraform generated `tf-networks.yml` file needs to use this prefix.

markgoddard commented on etc/kayobe/environments/ci-multinode/networks.yml:
```diff
@@ -8,87 +8,124 @@
+admin_oc_cidr: 10.209.0.0/16
+admin_oc_allocation_pool_start: 0.0.0.0
+admin_oc_allocation_pool_end: 0.0.0.0
+admin_oc_bootproto: dhcp
```
This needs to live in tf-networks.yml, since it depends on your cluster

jackhodgkiss replied:
Moved into `tf-networks.yml`

markgoddard commented on etc/kayobe/environments/ci-multinode/networks.yml:
```diff
@@ -8,87 +8,124 @@
+# Provision overcloud network
+provision_oc_cidr: 192.168.33.0/24
+provision_oc_mtu: 1450
+provision_oc_inspection_allocation_pool_start: 192.168.33.128
+provision_oc_inspection_allocation_pool_end: 192.168.33.254
```
No inspection required here. Should include a regular allocation pool though.

markgoddard commented on etc/kayobe/environments/ci-multinode/networks.yml:
```diff
@@ -8,87 +8,124 @@
+external_mtu: 1400
+external_allocation_pool_start: 192.168.38.3
+external_allocation_pool_end: 192.168.38.128
+external_vip_address: 192.168.38.2
```
VIP address not required for external

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/controllers/network-interfaces:
```diff
@@ -2,12 +2,17 @@
-aio_bridge_ports: []
+internal_interface: "{{ vxlan_interfaces[0].device }}.{{ internal_vlan }}"
+
+# external_interface: "{{ vxlan_interfaces[0].device }}.{{ external_vlan }}"
```
Can be removed

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/seed/network-interfaces:
```diff
@@ -1,13 +1,15 @@
+
+provision_oc_interface: "{{ vxlan_interfaces[0].device }}"
+
+external_interface: "{{ vxlan_interfaces[0].device }}.{{ external_vlan }}"
+
+public_interface: "{{ vxlan_interfaces[0].device }}.{{ public_vlan }}"
```
I think you are missing the seed.yml file that adds these two non-default networks.

Any other uncommitted files?

markgoddard commented:
Docs also need an update for deploying the seed and any other changes necessary

jackhodgkiss committed:
feat: add missing config files

jackhodgkiss committed:
fix: return admin interface to cephadm

jackhodgkiss committed:
feat: remove admin network config and introspection

g0rgamesh committed:
cherry-pick: Merge pull request #201 from stackhpc/doc-fix

add git to installed packages

jackhodgkiss committed:
feat: move VXLAN configuration under `ALL`

jackhodgkiss committed:
feat: add `provision_oc_vlan` and remove and `external_interface`

jackhodgkiss committed:
feat: enable neutron provider network

jackhodgkiss committed:
feat: override keepalived conf to increase `advert_int`

jackhodgkiss committed:
feat: add initial instructions for multiple network

jackhodgkiss committed:
feat: remove instructions about adding post hook

jackhodgkiss committed:
feat: update tempest instructions

jackhodgkiss committed:
feat: add `tips & tricks`

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/all/vxlan.yml:
```diff
@@ -1,6 +1,7 @@
 vxlan_interfaces:
-  - device: vxlan10
+  - device: "vxlan{{ vxlan_vni }}"
     group: 224.0.0.10
```
Could we use VNI here also? Maybe something with next_nth_usable from https://docs.ansible.com/ansible/latest/collections/ansible/utils/docsite/filters_ipaddr.html

jackhodgkiss replied:
Looking at https://infohub.delltechnologies.com/l/planning-guide-vmware-cloud-foundation-3-x-on-vxrail/select-a-multicast-ip-address-range-for-vxlan-network It appears it would be to also change 224.0.0.0 to 239.0.0.0.

Could add something such as `group: {{ '239.0.0.0/8' | ansible.utils.next_nth_usable(vxlan_vni) }}`

markgoddard commented on etc/kayobe/environments/ci-multinode/kolla/config/keepalived/keepalived.conf:
```diff
@@ -0,0 +1,48 @@
+    interface {{ api_interface }}
+    virtual_router_id {{ keepalived_virtual_router_id }}
+    priority {{ groups['haproxy'].index(inventory_hostname) + 1 }}
+    advert_int 3 # Increase advert interval from 1 to 3 as there was a lot of jumping amongst controllers on the internal network 
```
Did it fix the problem?

markgoddard commented on etc/kayobe/environments/ci-multinode/networks.yml:
```diff
@@ -98,6 +98,7 @@ provision_oc_cidr: 192.168.33.0/24
 provision_oc_mtu: 1450
 provision_oc_allocation_pool_start: 192.168.33.128
 provision_oc_allocation_pool_end: 192.168.33.254
+provision_oc_vlan: 191
```
Why not 107 (or the next free ID)?

jackhodgkiss replied:
Just picked a random ID as the `provision_oc` isn't being used. Will change it back to 107 to be more consistent.

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -1,48 +1,283 @@
 # Multinode Test Environment 
```
This cherry pick of a merge commit is a bit weird. It would have been better to merge in the multinode-env branch into multiple-networks to pull in the docs changes.

jackhodgkiss replied:
So I should have done

```
git checkout multiple-networks
git merge multinode-env
```

markgoddard commented on etc/kayobe/environments/ci-multinode/storage.yml:
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
git merge conflict artifact

markgoddard commented on etc/kayobe/hooks/overcloud-host-configure/pre.d/40-growroot.yml:
```diff
@@ -0,0 +1 @@
+/home/centos/src/kayobe-config/etc/kayobe/ansible/growroot.yml
```
I thought we weren't committing these? In any case they shouldn't use an absolute path.

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -1,48 +1,293 @@
+admin_allocation_pool_start: 0.0.0.0
+admin_allocation_pool_end: 0.0.0.0
+admin_bootproto: dhcp
+admin_ips:
```
admin_oc

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -1,48 +1,293 @@
+  kayobe-controller-03: 10.209.2.228
+```
+
+3. Configure the VXLAN interface for the `ALL` group 
````
nit: lowercase all

markgoddard replied:
TODO

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -1,48 +1,293 @@
+  vxlan_vni: 10
+  vxlan_interfaces:
+    - device: "vxlan{{ vxlan_vni }}"
+      group: 224.0.0.10
```
Leading space

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -1,48 +1,293 @@
+```
+
+3. Configure the VXLAN interface for the `ALL` group 
+{KAYOBE_CONFIG_PATH}/environments/${KAYOBE_ENVIRONMENT}/inventory/groups_vars/all/vxlan [See role documentation for more details](https://github.com/stackhpc/ansible-role-vxlan)
````
Let's just be explicit about what needs to be done here. It doesn't necessarily need to include a copy of the file (which might go stale).

markgoddard replied:
To be clear - we don't need the example (which could go stale) - just state that the vxlan_vni variable needs to be modified and must be unique within a network.

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -1,48 +1,293 @@
+kayobe control host bootstrap
+```
+
+2. Perform a overcloud host and seed configure
````
seed and overcloud host configure

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -1,48 +1,293 @@
+```
+
+```
+cp .automation.conf/tempest/tempest-{ci-aio,$KAYOBE_ENVIRONMENT}.overrides.conf
````
We should just commit this config

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -1,48 +1,293 @@
+In some instances the `configure-aio-resources.yml` may provide you with enough to work with and pass tempest tests. This can be achieved with the following;
+
+```
+kayobe playbook run etc/kayobe/ansible/configure-aio-resources.yml
````
This doesn't work for multinode, may as well remove it

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -1,48 +1,293 @@
+2. Ensure the private key for kayobe has been set
+
+```
+export KAYOBE_AUTOMATION_SSH_PRIVATE_KEY=$(cat ~/.ssh/id_rsa)
````
Makes me realise another issue with having terraform deploy the ansible control host - it needs the private key.

jackhodgkiss replied:
This has been handled with the deploy-openstack-config.yml playbook. https://github.com/stackhpc/terraform-kayobe-multinode/pull/7

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -1,48 +1,293 @@
+export KAYOBE_AUTOMATION_SSH_PRIVATE_KEY=$(cat ~/.ssh/id_rsa)
+```
+
+3. Update your tempest inventory file with your seed hostname
````
This can now be done in a standard way by adding the seed group as a child group of the tempest runner group. We can commit that file and remove this step.

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -1,48 +1,293 @@
+
+This section will be used to list useful commands and software that can be used to investigate and debug the mutltinode environment
+
+- When running the Tempest tests you may want to view the current progress which can be achieved by running `sudo docker logs --follow ${rally_docker_container_name}`
```
This could just go in the tempest section

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -1,48 +1,293 @@
+
+- When running the Tempest tests you may want to view the current progress which can be achieved by running `sudo docker logs --follow ${rally_docker_container_name}`
+
+- If need to access the network from your local machine or Ansible Control Host you can use `sshuttle` which operates as a transparent proxy over SSH. For example to access Horizon on the public network you can `sshuttle -r centos@${controller_ip} ${public_network_cidr}` This can also be used with `openstackclient`.
```
Let's make a new section before the tempest one about accessing and using the cloud.

Also, we know the public CIDR so we could include that in the command. For SMS lab is there a bit more to it than this? The Ansible control host typically won't have a public IP

jackhodgkiss committed:
feat: remove conflict artifact

jackhodgkiss committed:
feat: add docker logical volume

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/all/stackhpc-lvm.yml:
```diff
@@ -32,7 +32,8 @@ stackhpc_lvm_group_rootvg_lvs_default:
   - "{{ stackhpc_lvm_lv_home }}"
```
This file should contain the defaults, with overridden values in group_vars/controllers etc.

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -77,19 +77,6 @@ cd ../stackhpc-kayobe-config
 source kayobe-env --environment ci-multinode
 ```
 
-6. Add hooks for `configure-vxlan.yml` and `growroot.yml`
````
The hooks should not be committed to the repository - these instructions should stay

markgoddard replied:
These instructions should stay.

jackhodgkiss committed:
feat: update lvm and network interfaces

jackhodgkiss committed:
feat: add playbook/hooks for fixing network/hostname

markgoddard commented on etc/kayobe/hooks/overcloud-host-configure/pre.d/45-fix-hostname.yml:
```diff
@@ -0,0 +1 @@
+/home/centos/src/kayobe-config/etc/kayobe/ansible/fix-hostname.yml
```
As mentioned before - hooks should not be committed

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/all/stackhpc-lvm.yml:
```diff
@@ -33,16 +33,16 @@ stackhpc_lvm_group_rootvg_lvs_default:
 
```
This file should not be modified. It should be the same as etc/kayobe/inventory/group_vars/all/stackhpc-lvm in stackhpc/xena

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/storage/network-interfaces:
```diff
@@ -2,7 +2,7 @@
 # Network interface definitions for the storage group.
 
-admin_oc_interface: "eth0"
+admin_oc_interface: "ens3"
```
Greg already made this change - where did it go?

jackhodgkiss replied:
I think the problem here is the branch I have been working hasn't kept in sync with the `multinode-env` so changes like this have been repeated.

jackhodgkiss committed:
fix: revert changes back to stackhpc/xena

jackhodgkiss committed:
feat: add comments and change group address

jackhodgkiss committed:
feat: add tempest configuration for multinode environment

jackhodgkiss committed:
feat: add `ansible.utils` as requirement

`next_nth_usuable` requires this collection.

jackhodgkiss commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/seed/network-interfaces:
```diff
@@ -1,13 +1,15 @@
+provision_oc_interface: "{{ vxlan_interfaces[0].device}}.{{ provision_oc_vlan }}"
+
+external_interface: "{{ vxlan_interfaces[0].device }}.{{ external_vlan }}"
+
+public_interface: "{{ vxlan_interfaces[0].device }}.{{ public_vlan }}"
 
```
This seed.yml can found here: https://github.dev/stackhpc/stackhpc-kayobe-config/blob/b23cd49f142e77771995b700836cdf6f27ae0e76/etc/kayobe/environments/ci-multinode/seed.yml#L1

jackhodgkiss committed:
Merge branch 'multinode-env' into multiple-networks

jackhodgkiss committed:
fix: remove unwanted hooks

jackhodgkiss committed:
feat: update README

README now contains some additional information regarding openstack configure, VXLANs and tempest

markgoddard commented on .automation.conf/tempest/load-lists/tempest-full-skiplist:
```diff
@@ -0,0 +1 @@
+tempest.scenario.test_network_basic_ops.TestNetworkBasicOps.test_mtu_sized_frames.*: "Default MTU for tenant networks is greater than MTU on external network"
```
Unsure if it applies in our case

markgoddard commented on .automation.conf/tempest/tempest-ci-multinode.overrides.conf:
```diff
@@ -0,0 +1,29 @@
+
+[compute]
+# FIXME: set to 2 when we have multiple hypervisors
+min_compute_nodes = 1
```
You do have multiple hypervisors don't you?

jackhodgkiss replied:
Fixed.

markgoddard commented on .automation.conf/tempest/tempest-ci-multinode.overrides.conf:
```diff
@@ -0,0 +1,29 @@
+[compute-feature-enabled]
+resize = true
+# FIXME: enable when we have multiple hypervisors
+live_migration = false
```
Should it be true?

markgoddard commented on .automation.conf/tempest/tempest-ci-multinode.overrides.conf:
```diff
@@ -0,0 +1,29 @@
+resize = true
+# FIXME: enable when we have multiple hypervisors
+live_migration = false
+block_migration_for_live_migration = True
```
and this false?

jackhodgkiss replied:
Hadn't pushed the local changes.

markgoddard commented on etc/kayobe/ansible/requirements.yml:
```diff
@@ -6,5 +6,7 @@ collections:
   - name: stackhpc.cephadm
     version: 1.8.0
+  - name: ansible.utils
+    version: 2.6.1
```
I wonder if it works just using next_nth_usable without the FQCN?

jackhodgkiss replied:
It does work without the FQCN. Removed the `ansible.utils` requirement.

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/all/stackhpc-lvm.yml:
```diff
@@ -147,4 +147,4 @@ stackhpc_lvm_lv_docker:
   create: true
```
Probably need to move the changes you made in here into other group_vars files?

jackhodgkiss committed:
feat: move tempest skiplist in own dir and tweak overrides

jackhodgkiss committed:
feat: remove FQCN `ansible.utils`

jackhodgkiss committed:
feat: restore instructions on symlink hooks

jackhodgkiss committed:
feat: add newline back to `stackhpc-lvm.yml`

jackhodgkiss committed:
feat: add lvm variables for each group

jackhodgkiss committed:
fix: change `provision_oc_vlan` from 191 to 107

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -77,6 +77,21 @@ cd ../stackhpc-kayobe-config
+ln -s ${KAYOBE_CONFIG_PATH}/ansible/growroot.yml 5-growroot.yml
+ln -s ${KAYOBE_CONFIG_PATH}/ansible/fix-hostname.yml 10-fix-hostname.yml
+ln -s ${KAYOBE_CONFIG_PATH}/ansible/fix-networking.yml 15-fix-networking.yml
+ln -s ${KAYOBE_CONFIG_PATH}/ansible/configure-vxlan.yml 20-configure-vxlan.yml
```
`cd ..`

jackhodgkiss replied:
Could you clarify the comment please. Running the above set of commands provides the desired result with all symlinks be defined where they should be.

markgoddard replied:
ok, yes you're right - it's using an absolute path.

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/compute/lvm.yml:
```diff
@@ -0,0 +1,33 @@
+# List of extra LVs to include in the rootvg VG.
+stackhpc_lvm_group_rootvg_lvs_extra:
+  - stackhpc_lvm_lv_docker
```
Needs `"{{ }}"`

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/controllers/lvm.yml:
```diff
@@ -0,0 +1,33 @@
+# List of extra LVs to include in the rootvg VG.
+stackhpc_lvm_group_rootvg_lvs_extra:
+  - stackhpc_lvm_lv_docker
```
ditto

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/seed/lvm.yml:
```diff
@@ -0,0 +1,33 @@
+# List of extra LVs to include in the rootvg VG.
+stackhpc_lvm_group_rootvg_lvs_extra:
+  - stackhpc_lvm_lv_docker
```
ditto

markgoddard commented on etc/kayobe/environments/ci-multinode/inventory/group_vars/storage/lvm.yml:
```diff
@@ -1,3 +1,10 @@
+# List of extra LVs to include in the rootvg VG.
+stackhpc_lvm_group_rootvg_lvs_extra:
+  - stackhpc_lvm_lv_docker
```
ditto

markgoddard commented on .automation.conf/tempest/tempest-ci-multinode.overrides.conf:
```diff
@@ -0,0 +1,32 @@
+
+[compute-feature-enabled]
+resize = true
+# FIXME: enable when we have multiple hypervisors
```
Stale comment

markgoddard commented on .automation.conf/tempest/tempest-ci-multinode.overrides.conf:
```diff
@@ -0,0 +1,32 @@
+v3_endpoint_type = publicURL
+
+[compute]
+# FIXME: set to 2 when we have multiple hypervisors
```
Stale comment

jackhodgkiss committed:
fix: syntax issue "{{ }}"

jackhodgkiss committed:
fix: remove stale comments

jackhodgkiss committed:
fix: vxlan instructions

jackhodgkiss committed:
fix: rename `admin_` to `admin_oc_`

jackhodgkiss committed:
fix: change vxlan_vni range

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -77,17 +77,19 @@ cd ../stackhpc-kayobe-config
+ln -s ${KAYOBE_CONFIG_PATH}/ansible/fix-networking.yml 15-fix-networking.yml
+ln -s ${KAYOBE_CONFIG_PATH}/ansible/configure-vxlan.yml 20-configure-vxlan.yml
+cd ${KAYOBE_CONFIG_PATH}/hooks/seed-host-configure/pre.d
+ln -s ${KAYOBE_CONFIG_PATH}/ansible/growroot.yml 5-growroot.yml
```
Surprised that the seed doesn't need fix-networking.yml

jackhodgkiss replied:
Good spot. Missed the symlink when writing the docs. Added fix-networking to seed-host-configure.

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -101,8 +103,10 @@ kayobe-controller-03
 
 [compute]
 kayobe-compute-01
```
Inventory comes from terraform now

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
```diff
@@ -114,53 +118,55 @@ osds
-  kayobe-ceph-1: 10.209.0.76
-  kayobe-ceph-2: 10.209.3.225
-  kayobe-ceph-3: 10.209.1.20
+admin_oc_cidr: 10.209.0.0/16
```
This comes from Terraform now

jackhodgkiss committed:
fix: remove inventory and admin_oc from docs

jackhodgkiss committed:
fix: add missing symlink from docs

markgoddard commented on etc/kayobe/environments/ci-multinode/README.md:
````diff
@@ -89,68 +89,12 @@ ln -s ${KAYOBE_CONFIG_PATH}/ansible/fix-networking.yml 15-fix-networking.yml
-  kayobe-controller-03: 10.209.2.228
-```
-
 3. Configure the VXLAN interface for the `all` group
````
1.

jackhodgkiss committed:
fix: reset list numbered list

markgoddard reviewed (approved):
I think let's merge this as it is.

Looking at the README it doesn't quite make sense with the new ansible playbook in the terraform repo, and some of the workarounds are now automated as hooks. Unsure whether we should assume the use of the playbook in this README, or document all steps for completeness.
